generator/bench/Bench/Pos/Criterion/Block/Logic.hs

```diff
@@ -0,0 +1,243 @@
+        let epoch = eos ^. epochIndexL
+        let secretKeys = case genesisSecretKeys of
+                Nothing -> error "verifyHeaderBench: no genesisSecretKeys"
+                Just ks -> ks
+        let blockGenParams = BlockGenParams
+                { _bgpSecrets = mkAllSecretsSimple secretKeys
+                , _bgpBlockCount = BlockCount 1
+                , _bgpTxGenParams = TxGenParams
+                    { _tgpTxCountRange = (0, 2)
+                    , _tgpMaxOutputs = 2
+                    }
+                , _bgpInplaceDB = False
+                , _bgpSkipNoKey = True -- TODO: should be False?
+                , _bgpGenStakeholders = gdBootStakeholders genesisData
+                , _bgpTxpGlobalSettings = txpGlobalSettings pm
+                }
+        leaders <- lrcActionOnEpochReason epoch "genBlock" LrcDB.getLeadersForEpoch
+        mblock <- flip evalRandT g $ do
+            blockGenCtx <- lift $ mkBlockGenContext blockGenParams
+            tipHeader <- lift $ getTipHeader
+            mapRandT (flip runReaderT blockGenCtx)
+                $ genBlockNoApply pm eos tipHeader
+        let !block = fromMaybe (error "verifyHeaderBench: failed to generate a header") mblock
+
+        let !verifyHeaderParams = VerifyHeaderParams
+                { vhpPrevHeader = Nothing
+                , vhpCurrentSlot = Nothing
+                , vhpLeaders = Just leaders
+                , vhpMaxSize = Nothing
+                , vhpVerifyNoUnknown = False
+                }
+        let !params = VerifyBlockParams
+                { vbpVerifyHeader = verifyHeaderParams
+                , vbpMaxSize = bvdMaxBlockSize (_tpBlockVersionData tp)
+                , vbpVerifyNoUnknown = False
+                }
+
+        return (block, params)
+
+    benchHeaderVerification ~(header, params) =
+        nf isVerSuccess $ verifyHeader pm params header
+
+    benchBlockVerification ~(block, params) =
+        nf isVerSuccess $ verifyBlock pm params block
+
+
+runBenchmark :: IO ()
+runBenchmark = do
+    startTime <- realTime
+    let cfo = defaultConfigurationOptions
+            { cfoFilePath = "../lib/configuration.yaml"
+            , cfoKey = "bench-validation"
+            , cfoSystemStart = Just (Timestamp startTime)
+            }
+    withCompileInfo $
+        withConfigurationsM (LoggerName "verifyBenchmark") Nothing cfo $ \_ pm -> do
+                let tp = TestParams
+                        { _tpStartTime = Timestamp (convertUnit startTime)
+                        , _tpBlockVersionData = genesisBlockVersionData
+                        , _tpGenesisInitializer = genesisInitializer
+                        }
+                runEmulation startTime
+                    $ initBlockTestContext tp $ \ctx ->
+                        sudoLiftIO $ defaultMainWith config
+                            [ verifyBlocksBenchmark pm tp ctx
+                            , verifyHeaderBenchmark pm tp
```

Comment:
Do these write to / read from disk? It's hard to tell when this stuff is determined by the monad.

Reply:
No, they don't.  The block cache is only implemented for the profiling tool `cardano-sl-verification-bench-exe`.